### PR TITLE
Add skeleton code for reconcile service

### PR DESCRIPTION
**Pull request by merlintang**
This PR adds skeleton code for reconcile service.

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/common/25)

### CODE REVIEW

**Comment by merlintang** [comment]
I signed it!

File: job_controller/job_controller.go [hunk 1]
```diff
@@ -87,7 +87,10 @@ type ControllerInterface interface {
 	// DeletePod deletes the pod
 	DeletePod(job interface{}, pod *v1.Pod) error
 
-	// SetClusterSpec sets the cluster spec for the pod
+	// Get the deafult container port number
+	GetDefaultContainerPortName() string
```
Comment (gaocegege) [inline job_controller/job_controller.go]: Could we place the function after `GetDefaultContainerName() string`?

File: job_controller/service.go [hunk 2]
```diff
@@ -157,3 +162,113 @@ func (jc *JobController) GetServiceSlices(services []*v1.Service, replicas int,
...
+			err = jc.createNewService(job, rtype, spec, strconv.Itoa(index))
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// GetPortFromJob gets the port of tensorflow container.
```
Comment (terrytangyuan) [inline job_controller/service.go]: Remove "tensorflow"

File: job_controller/service.go [hunk 3]
```diff
@@ -157,3 +162,113 @@ func (jc *JobController) GetServiceSlices(services []*v1.Service, replicas int,
...
+			}
+		}
+	}
+	return -1, fmt.Errorf("failed to found the port")
+}
+
+// createNewService creates a new service for the given index and type.
+func (jc *JobController) createNewService(job metav1.Object, rtype commonv1.ReplicaType,
+	spec *commonv1.ReplicaSpec, index string) error {
+	tfjobKey, err := KeyFunc(job)
```
Comment (terrytangyuan) [inline job_controller/service.go]: Remove "tf"

File: job_controller/job_controller.go [hunk 4]
```diff
@@ -93,6 +93,9 @@ type ControllerInterface interface {
 	// Returns the default container name in pod
 	GetDefaultContainerName() string
 
+	// Get the deafult container port number
```
Comment (terrytangyuan) [inline job_controller/job_controller.go]: port number -> port name to be consistent with the function name

**Comment by terrytangyuan** [comment]
@merlintang Travis CI build [failed](https://travis-ci.com/kubeflow/common/builds/109397731) on `golangci-lint run`. I believe you can fix it by changing `reconcileServices` to `ReconcileServices` for now.

File: job_controller/service.go [hunk 5]
```diff
@@ -157,3 +162,113 @@ func (jc *JobController) GetServiceSlices(services []*v1.Service, replicas int,
...
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// GetPortFromJob gets the port of job container.
+func (jc *JobController) GetPortFromJob(spec *commonv1.ReplicaSpec) (int32, error) {
```
[the remark this answers is not in the document]
Comment (merlintang) [reply job_controller/service.go]: the input is the replicaSpec, then the underline logic is the replicaSpec related container. thus, I prefer the container name for further understand. if you think replicaSpec is better for understanding, I can change it.

File: job_controller/service.go [hunk 6]
```diff
@@ -157,3 +162,113 @@ func (jc *JobController) GetServiceSlices(services []*v1.Service, replicas int,
...
+		if container.Name == jc.Controller.GetDefaultContainerName() {
+			ports := container.Ports
+			for _, port := range ports {
+				if port.Name == jc.Controller.GetDefaultContainerPortNumber(){
+					return port.ContainerPort, nil
+				}
+			}
+		}
+	}
+	return -1, fmt.Errorf("failed to found the port")
```
[the remark this answers is not in the document]
Comment (merlintang) [reply job_controller/service.go]: thanks! Liu, I have updated them.

File: job_controller/job_controller.go [hunk 7]
```diff
@@ -48,6 +48,12 @@ type ControllerInterface interface {
 	// Returns the Group Name(value) in the labels of the job
 	GetGroupNameLabelValue() string
 
+	// Returns the Replica Type(key) in the labels of the job
+	GetReplicaTypeLabelKey() string
```
Comment (johnugeorge) [inline job_controller/job_controller.go]: Is this because of merge conflict? https://github.com/kubeflow/common/pull/29 removes these labels

Comment (terrytangyuan) [reply job_controller/job_controller.go]: I think so. @merlintang Could you rebase?

Comment (merlintang) [reply job_controller/job_controller.go]: it is fixed. thanks for your suggestion.

**Comment by merlintang** [comment]
@jian-he  can you review this as well?

File: job_controller/job_controller.go (hunk 7 above)
Comment (jian-he) [reply job_controller/job_controller.go]: These interfaces are supposed to be removed and use [this](https://github.com/kubeflow/common/blob/39cbea8d8da2c633a706143f03fee442fa0c1f10/util/status.go#L22) instead

File: job_controller/job_controller.go [hunk 8]
```diff
@@ -49,6 +49,12 @@ type ControllerInterface interface {
...
+
+	// Returns the Replica Index(value) in the labels of the job
+	GetReplicaIndexLabelKey() string
+
```
Comment (jian-he) [inline job_controller/job_controller.go]: same for this

File: job_controller/test_job_controller.go [hunk 9]
```diff
@@ -32,10 +32,22 @@ func (TestJobController) GetGroupNameLabelValue() string {
...
 }
 
+func (TestJobController) GetReplicaTypeLabelKey() string {
+	return util.ReplicaTypeLabel
+}
+
+func (TestJobController) GetReplicaIndexLabelKey() string {
+	return util.ReplicaIndexLabel
+}
+
```
Comment (jian-he) [inline job_controller/test_job_controller.go]: same remove these two

File: job_controller/job_controller.go (hunk 8 above)
Comment (jian-he) [reply job_controller/job_controller.go]: These two also need to be fixed.

https://github.com/kubeflow/common/blob/b615b1c764233de73d0a6f2d20d1a30bac6d45c6/job_controller/job.go#L206

https://github.com/kubeflow/common/blob/b615b1c764233de73d0a6f2d20d1a30bac6d45c6/job_controller/job.go#L103

File: job_controller/job.go [hunk 10]
```diff
@@ -204,12 +203,12 @@ func (jc *JobController) ReconcileJobs(
 		}
 
 		// TODO(terrytangyuan): Uncomment this once service reconciliation logic is in place
```
Comment (jian-he) [inline job_controller/job.go]: could you remove this TODO ?

**Comment by jian-he** [comment]
@richardsliu take a look ?

**Pull request merged**